### ISSUE TITLE
Memory error for multiple dataloaders

### ISSUE THREAD

**Issue by cyk1337** [huggingface/accelerate#652]
For the use of multiple datasets and tasks, we use around more than 200+ dataloaders, then pass it into `dataloader1, dataloader2, ..., dataloader200=accelerate.prepare(dataloader1, dataloader2, ..., dataloader200)`
It causes the memory error when generating batches. Any solutions to it?

```bash
  File "/home/xxx/my_code/src/utils/data_utils.py", line 54, in generate_batch
    x = next(iterator)
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/accelerate/data_loader.py", line 301, in __iter__
    for batch in super().__iter__():
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 435, in __next__
    data = self._next_data()
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 475, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 28, in fetch
    data.append(next(self.dataset_iter))
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/accelerate/data_loader.py", line 249, in __iter__
    for element in self.dataset:
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/datasets/iterable_dataset.py", line 503, in __iter__
    for key, example in self._iter():
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/datasets/iterable_dataset.py", line 500, in _iter
    yield from ex_iterable
  File "/home/xxx/anaconda3/envs/pt1.7/lib/python3.7/site-packages/datasets/iterable_dataset.py", line 231, in __iter__
    new_key = "_".join(str(key) for key in keys)
MemoryError
```

**Comment by pacman100** [comment]
Hello @cyk1337, is this the same issue related to #653?

Also, can you provide a minimal script to reproduce the issue?

**Comment by cyk1337** [comment]
Yeah, same issue, but different errors. 

Generally, I used 100+ custom datasets, of each has a (torch) dataloader. Then pass then into the `accelerate.prepare` for distributed settings. 
1. If passing the dataloader into  `accelerate.prepare`, the training loop will get hang after very few steps.
2. If not, it would cause the above `MemoryError ` (OOM). I guess that due to the prefeching of each dataloader, the cost of 100+ dataloaders would be high. But unsure of how to locate and debug it.

Any possible solutions?

As to the minimal script, I am trying to reproduce one using publica datasets, still on the way.

**Comment by pacman100** [comment]
Hello, I think the issue is the design choices. Can't you concatenate 100 datasets into a single big one and then use that?

**Comment by cyk1337** [comment]
> Hello, I think the issue is the design choices. Can't you concatenate 100 datasets into a single big one and then use that?

It is to reproduce the sampling strategy for low-resource languages as in XLM(-R). Using which dataloader is sampled rather than equiprobable. Another issue is, I reproduced T5 span mask strategy, which will concatenate adjacent samples within a batch. If concatenating and shuffling re-distributed samples, the mask span may be across different datasources, like different languages. This is why I use multiple dataloaders as in XLM.
